File: Exercises/T3/src/java/GestaoABC/Funcionaria.java
package Exercises.T3.src.java.GestaoABC;
import Exercises.T3.src.java.Validacoes.CpfInvalidoException;

public abstract class Funcionaria {
  public String nome;
  public String CPF;
  public double salarioBase;

  private static int obterRestoCpfTruncado(String CPF, int posicaoTruncamento) {
    int multiplicador = posicaoTruncamento;
    int soma = 0;
    for (int i = 0; i < posicaoTruncamento-1; i++) {
      int cpfNumI = Character.getNumericValue(CPF.charAt(i));
      soma += (multiplicador * cpfNumI);
      multiplicador--;
    }

    
    int resto = (soma * 10) % 11;
    if (resto == 10 || resto == 11) {resto = 0;}
    return resto;
  }

  public static void verificaCPF(String CPF) throws CpfInvalidoException {
    if (CPF.length() != 11) {
      throw new CpfInvalidoException(
        String.format("Tamanho inválido de CPF.%nInformado: %s%n", CPF)
      );
    };

    int resto = obterRestoCpfTruncado(CPF, 10);

    int digitoVerificador = Character.getNumericValue(CPF.charAt(9));
    if (resto != digitoVerificador) {
      throw new CpfInvalidoException(
        String.format("Número inválido de CPF.%nInformado: %s%n", CPF)
      );
    }

    resto = obterRestoCpfTruncado(CPF, 11);

    digitoVerificador = Character.getNumericValue(CPF.charAt(10));
    if (resto != digitoVerificador) {
      throw new CpfInvalidoException(
        String.format("Número inválido de CPF.%nInformado: %s%n", CPF)
      );
    }
  }

  public String superToString() {
    String strF = String.format(
      "- Nome: %s%n- CPF: %s%n- Salario base: %,.2f%n",
      this.nome, this.CPF, this.salarioBase
    );
    return strF;
  }

  public abstract double calculaSalario();
}

File: Exercises/T3/src/python/GestaoABC/Funcionaria.py
from abc import ABC, abstractmethod

class CpfInvalido(Exception):
    """
    Erro levantado caso CPF informado invalido
    """
    pass

class Funcionaria(ABC):
    def __init__(self, nome: str, CPF: str, salarioBase: float) -> None:
        self.nome = nome
        self.CPF = CPF
        self.salarioBase = salarioBase

    @abstractmethod
    def calculaSalario() -> float:
        pass
    
    @staticmethod
    def obterRestoCpfTruncado(CPF: str, posicaoTruncamento: int) -> int:
        multiplicador = posicaoTruncamento
        soma = 0
        for i in range(posicaoTruncamento-1):
            cpfNumI = int(CPF[i])
            soma += (multiplicador * cpfNumI)
            multiplicador -= 1

        resto = (soma * 10) % 11
        if resto == 10 or resto == 11:
            resto = 0
        return resto

    @staticmethod
    def verificaCPF(CPF: str):
        if len(CPF) != 11: 
            raise CpfInvalido("Tamanho inválido de CPF. Informado: " + CPF)

        resto = Funcionaria.obterRestoCpfTruncado(CPF, 10)
        digitoVerificador = int(CPF[9])

        if resto != digitoVerificador:
            raise CpfInvalido("Número inválido de CPF. Informado: " + CPF)
        
        resto = Funcionaria.obterRestoCpfTruncado(CPF, 11)
        digitoVerificador = int(CPF[10])

        if resto != digitoVerificador:
            raise CpfInvalido("Número inválido de CPF. Informado: " + CPF)

    def __str__(self) -> str:
        return f" Nome: {self.nome}\n CPF: {self.CPF}\n Salario base: {self.salarioBase}\n"
    
    
